### PR TITLE
Portable code

### PR DESCRIPTION
**Pull request by loomis**
For review.  Probably should not be merged.

### CODE REVIEW

**Comment by loomis** [comment]
Connected to #12.

**Comment by konstan** [comment]
Tests in ClojureScript part require `phantomjs`, but this doesn't fail the build process.

```
14:34 $ boot build
...
We tried running phantomjs but we couldn't find it your system. Try:

         phantomjs -v
...
```

Full output on my machine.  Why does it (try to?) test `cats`, `kvlt` and `promesa` which are just dependencies?

```
14:34 $ boot build
Checking out default-deps-3.1-SNAPSHOT.jar...
Writing output.cljs.edn...
Writing clj_test/suite.cljs...
Compiling ClojureScript...
WARNING: Replacing ClojureScript compiler option :main with automatically set value.
• output.js
WARNING: Unknown compiler option ':warning-handlers'.
Running cljs tests...[doo] Failed to run command:  #error {
 :cause "error=2, No such file or directory"
 :via
 [{:type java.io.IOException
   :message "Cannot run program \"phantomjs\" (in directory \"/Users/konstan/.boot/cache/tmp/Users/konstan/SixSq/code/SlipStreamClientAPI/clj/1fsg/-p9hf6h\"): error=2, No such file or directory"
   :at [java.lang.ProcessBuilder start "ProcessBuilder.java" 1048]}
  {:type java.io.IOException
   :message "error=2, No such file or directory"
   :at [java.lang.UNIXProcess forkAndExec "UNIXProcess.java" -2]}]
 :trace
 [[java.lang.UNIXProcess forkAndExec "UNIXProcess.java" -2]
  [java.lang.UNIXProcess <init> "UNIXProcess.java" 248]
  [java.lang.ProcessImpl start "ProcessImpl.java" 134]
  [java.lang.ProcessBuilder start "ProcessBuilder.java" 1029]
  [java.lang.Runtime exec "Runtime.java" 620]
  [doo.shell$exec_BANG_ invokeStatic "shell.clj" 48]
  [doo.shell$exec_BANG_ invoke "shell.clj" 43]
  [doo.shell$sh invokeStatic "shell.clj" 75]
  [doo.shell$sh invoke "shell.clj" 70]
  [doo.core$run_script invokeStatic "core.clj" 264]
  [doo.core$run_script invoke "core.clj" 232]
  [clojure.lang.Var invoke "Var.java" 388]
  [crisptrutski.boot_cljs_test$eval1117$fn__1118$fn__1124$fn__1125 invoke "boot_cljs_test.clj" 99]
  [adzerk.boot_cljs$eval444$fn__445$fn__451$fn__452 invoke "boot_cljs.clj" 201]
  [adzerk.boot_cljs$eval393$fn__394$fn__400$fn__401 invoke "boot_cljs.clj" 135]
  [crisptrutski.boot_cljs_test$eval1078$fn__1079$fn__1085$fn__1086 invoke "boot_cljs_test.clj" 65]
  [crisptrutski.boot_cljs_test$capture_fileset$fn__1155$fn__1156 invoke "boot_cljs_test.clj" 110]
  [boot.core$run_tasks invoke "core.clj" 794]
  [boot.core$boot$fn__904 invoke "core.clj" 804]
  [clojure.core$binding_conveyor_fn$fn__4676 invoke "core.clj" 1938]
  [clojure.lang.AFn call "AFn.java" 18]
  [java.util.concurrent.FutureTask run "FutureTask.java" 266]
  [java.util.concurrent.ThreadPoolExecutor runWorker "ThreadPoolExecutor.java" 1142]
  [java.util.concurrent.ThreadPoolExecutor$Worker run "ThreadPoolExecutor.java" 617]
  [java.lang.Thread run "Thread.java" 745]]}
We tried running phantomjs but we couldn't find it your system. Try:

         phantomjs -v

If it doesn't work you need to install phantomjs, see https://github.com/bensu/doo#setting-up-environments

If it does work, file an issue and we'll sort it together!
Writing main.cljs.edn...
Compiling ClojureScript...
• main.js
WARNING: Unknown compiler option ':warning-handlers'.
Mar 23, 2016 2:35:44 PM com.google.javascript.jscomp.LoggerErrorManager println
WARNING: /Users/konstan/.boot/cache/tmp/Users/konstan/SixSq/code/SlipStreamClientAPI/clj/1fsg/avn0o5/main.out/promesa/core.js:69: WARNING - Keywords and reserved words are not allowed as unquoted property names in older versions of JavaScript. If you are targeting newer versions of JavaScript, set the appropriate language_in option.
return it__$1.catch(cb);
              ^

Mar 23, 2016 2:35:44 PM com.google.javascript.jscomp.LoggerErrorManager println
WARNING: /Users/konstan/.boot/cache/tmp/Users/konstan/SixSq/code/SlipStreamClientAPI/clj/1fsg/avn0o5/main.out/promesa/core.js:266: WARNING - Keywords and reserved words are not allowed as unquoted property names in older versions of JavaScript. If you are targeting newer versions of JavaScript, set the appropriate language_in option.
return p.finally(callback);
         ^

Mar 23, 2016 2:35:44 PM com.google.javascript.jscomp.LoggerErrorManager printSummary
WARNING: 0 error(s), 2 warning(s)
Writing pom.xml and pom.properties...

Testing com.sixsq.slipstream.clj-client.lib.context

Testing com.sixsq.slipstream.clj-client.lib.http-impl

Testing com.sixsq.slipstream.clj-client.lib.run

Testing com.sixsq.slipstream.clj-client.lib.run-impl

Testing com.sixsq.slipstream.clj-client.lib.utils

Testing com.sixsq.slipstream.clj-client.lib.utils.context

Testing com.sixsq.slipstream.clj-client.lib.utils.wait

Testing com.sixsq.slipstream.clj-client.run

Testing com.sixsq.slipstream.clj-client.run-impl-test

Testing com.sixsq.slipstream.clj-client.run-test

Testing com.sixsq.slipstream.clj-client.utils-test

Testing cats.context

Testing cats.core

Testing cats.protocols

Testing cats.util

Testing kvlt.core

Testing kvlt.event-source

Testing kvlt.middleware

Testing kvlt.middleware.params

Testing kvlt.middleware.util

Testing kvlt.util

Testing kvlt.websocket

Testing promesa.core

Testing promesa.protocols

Ran 8 tests containing 36 assertions.
0 failures, 0 errors.
Compiling 1/24 cats.context...
Compiling 2/24 cats.core...
Compiling 3/24 cats.protocols...
Compiling 4/24 cats.util...
Compiling 5/24 com.sixsq.slipstream.clj-client.lib.context...
Compiling 6/24 com.sixsq.slipstream.clj-client.lib.http-impl...
Compiling 7/24 com.sixsq.slipstream.clj-client.lib.run...
Compiling 8/24 com.sixsq.slipstream.clj-client.lib.run-impl...
Compiling 9/24 com.sixsq.slipstream.clj-client.lib.utils...
Compiling 10/24 com.sixsq.slipstream.clj-client.lib.utils.context...
Compiling 11/24 com.sixsq.slipstream.clj-client.lib.utils.wait...
Compiling 12/24 com.sixsq.slipstream.clj-client.run...
Compiling 13/24 com.sixsq.slipstream.clj-client.run-impl-test...
Compiling 14/24 com.sixsq.slipstream.clj-client.run-test...
Compiling 15/24 com.sixsq.slipstream.clj-client.utils-test...
Compiling 16/24 kvlt.core...
Compiling 17/24 kvlt.event-source...
Compiling 18/24 kvlt.middleware...
Compiling 19/24 kvlt.middleware.params...
Compiling 20/24 kvlt.middleware.util...
Compiling 21/24 kvlt.util...
Compiling 22/24 kvlt.websocket...
Compiling 23/24 promesa.core...
Compiling 24/24 promesa.protocols...
Writing SlipStreamClientAPI-jar-3.1-SNAPSHOT.jar...
Installing SlipStreamClientAPI-jar-3.1-SNAPSHOT.jar...
Writing target dir(s)...
14:36 $
```

**Comment by loomis** [comment]
Yes, forgot to mention that phantomjs is required.  On Mac OS X, it is available via `brew`.  Will have to add this to the contributor's guide and see about installing it on jenkins.

**Comment by konstan** [comment]
OK.  Thanks.  
What about tests of `cats`, `kvlt` and `promesa` ?

File: clj/build.boot [hunk 1]
```diff
@@ -38,9 +47,10 @@
           :title       "SlipStream Client API"
           :sources     #{"src"}
           :description "Client library to interact with SlipStream via REST API."
-          :target      "target/doc/api"})
+          :target      "target/doc/api"}
+  test-cljs {:js-env :phantom})
 
-(deftask build
+(deftask build-clj
          "build full project"
```
Comment (konstan) [inline clj/build.boot]: Should be changed to "build clojure part of the project".

```
15:21 $ boot build-clj
Checking out default-deps-3.1-SNAPSHOT.jar...
Writing pom.xml and pom.properties...

Testing com.sixsq.slipstream.clj-client.run-impl-test

Testing com.sixsq.slipstream.clj-client.run-test

Testing com.sixsq.slipstream.clj-client.utils-test

Testing com.sixsq.slipstream.clj-client.lib.context

Testing com.sixsq.slipstream.clj-client.lib.http-impl

Testing com.sixsq.slipstream.clj-client.lib.run

Testing com.sixsq.slipstream.clj-client.lib.run-impl

Testing com.sixsq.slipstream.clj-client.lib.utils

Testing com.sixsq.slipstream.clj-client.lib.utils.context

Testing com.sixsq.slipstream.clj-client.lib.utils.wait

Testing com.sixsq.slipstream.clj-client.run

Ran 8 tests containing 36 assertions.
0 failures, 0 errors.
Compiling 1/11 com.sixsq.slipstream.clj-client.lib.context...
Compiling 2/11 com.sixsq.slipstream.clj-client.lib.http-impl...
Compiling 3/11 com.sixsq.slipstream.clj-client.lib.run...
Compiling 4/11 com.sixsq.slipstream.clj-client.lib.run-impl...
Compiling 5/11 com.sixsq.slipstream.clj-client.lib.utils...
Compiling 6/11 com.sixsq.slipstream.clj-client.lib.utils.context...
Compiling 7/11 com.sixsq.slipstream.clj-client.lib.utils.wait...
Compiling 8/11 com.sixsq.slipstream.clj-client.run...
Compiling 9/11 com.sixsq.slipstream.clj-client.run-impl-test...
Compiling 10/11 com.sixsq.slipstream.clj-client.run-test...
Compiling 11/11 com.sixsq.slipstream.clj-client.utils-test...
Writing SlipStreamClientAPI-jar-3.1-SNAPSHOT.jar...
Installing SlipStreamClientAPI-jar-3.1-SNAPSHOT.jar...
Writing target dir(s)...
✔ ~/code/SlipStreamClientAPI/clj [portable-code|✚ 1⚑ 1]
15:22 $
```

**Comment by loomis** [comment]
Adding the flag `--exit? true` to the `test-cljs` task will have the build fail on testing errors.

**Comment by konstan** [comment]
You probably meant `:exit?`. And yes this works.

**Comment by konstan** [comment]
For me to understand.  What is an intended result of invoking `boot buil-cljs`?  Shouldn't it be some kind of artefact?  I see nothing produced in the end.  Thanks.

```
15:48 $ boot build-cljs
Checking out default-deps-3.1-SNAPSHOT.jar...
Writing output.cljs.edn...
Writing clj_test/suite.cljs...
Compiling ClojureScript...
WARNING: Replacing ClojureScript compiler option :main with automatically set value.
• output.js
WARNING: Unknown compiler option ':warning-handlers'.
Running cljs tests...
Testing com.sixsq.slipstream.clj-client.utils-test

Testing com.sixsq.slipstream.clj-client.run-impl-test

Ran 6 tests containing 31 assertions.
0 failures, 0 errors.
✔ ~/code/SlipStreamClientAPI/clj [portable-code|✚ 1⚑ 1]
15:48 $
```

**Comment by loomis** [comment]
At the moment, just a complete compilation of the code.  The artifact should be a source jar file for both clojure and clojurescript which would then be included/used by other things.  This could be done either by producing multiple artifacts or creating the uberjar based on the source jar file.

**Comment by loomis** [comment]
Unfortunately `phantomjs` is not easily installed on CentOS 7 (at least not via an RPM); however, `rhino` is available.  Need to test to see if the clojurescript tests work with rhino. Also need to see how easily `rhino` can be installed on the Mac.

File: clj/build.boot [hunk 2]
```diff
@@ -51,6 +62,16 @@
...
 
+(deftask build-cljs
+  "build clojurescript"
+  []
+  (comp
+   (test-cljs)
+   (cljs :optimizations :advanced)))
+
+(deftask build []
+  (comp (build-cljs) (build-clj)))
```
Comment (konstan) [inline clj/build.boot]: Unfortunately, the composition doesn't work in this order. `build-clj` is not invoked.  I didn't investigate why. Suggestion is to change to:

``` clojure
(deftask build-clj
         "build clojure"
         []
         (comp
           (pom)
           (test)
           (aot :all true)
           (jar)
           (install)
           (target)))

(deftask build-cljs
         "build clojurescript"
         []
         (comp
           (test-cljs)
           (cljs :optimizations :advanced)))

(deftask build
         []
         "build full project"
         (comp
           (build-clj)
           (build-cljs)))
```

In which case both sub-tasks get built.

Comment (loomis) [inline clj/build.boot]: It seems to be a result of setting `:exit?` to true.  It seems to exit whether or not there are errors.  Both are built if `:exit?` is false.

Comment (konstan) [inline clj/build.boot]: But as I understand we need `:exit? true` to exit if there are failed tests.  Simple workaround would be just to change the order of the tasks.

Comment (loomis) [inline clj/build.boot]: We'd also need to switch the order of `(test-cljs)` and `(cljs. ...)` because the compilation is also not run.

**Comment by loomis** [comment]
Did the simple workaround for now.  Moved the cljs build to the end.  The ClojureScript build is not done when building from maven to avoid interference with deployment of artifacts.

**Comment by loomis** [comment]
The only "easy" javascript environment that works with `boot/doo` is `phantomjs`.  Others are supposed to work, but I've not been able to get them to run correctly.  Installing `phantomjs` is easy on Mac OS X via `brew`.  On linux it requires downloading a tarball and putting the executable in the standard path.

**Comment by loomis** [comment]
Installation of phantomjs has been added to the draft documentation and to the CI setup of the build environment.

**Pull request merged**